### PR TITLE
Connect StakePool.Metrics with NetworkLayer and Api.Server

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#711 

# Overview

- [x] Commit 1: I made the constructors for Jörmungandr `NetworkLayer`s use raw blocks (`Jormungandr.Binary.Block`) that are later converted to Wallet- and StakePoolMetrics-blocks.
- [x] Commit 2: I simplified the types of `JormungandrClient.getStakeDistribution` and added `NetworkLayer.stakeDistribution` since our code currently works well with a large `NetworkLayer`.
- [x] Commit 3: I added (untested) function for merging the stake-distribution and stake-pool metrics.
- [x] Commit 3: I added `StakePoolLayer { listStakePools }` and connected this to the Server along with some error messages.

# Comments

```bash
$ cardano-wallet-jormungandr stake-pool list
Ok.
[
    {
        "metrics": {
            "controlled_stake": {
                "quantity": 0,
                "unit": "lovelace"
            },
            "produced_blocks": {
                "quantity": 12,
                "unit": "block"
            }
        },
        "id": "0ec3c1f81e80038d360b890c346956c85c918e2aca5b28d12d802198acab7175"
    },
    {
        "metrics": {
            "controlled_stake": {
                "quantity": 0,
                "unit": "lovelace"
            },
            "produced_blocks": {
                "quantity": 79,
                "unit": "block"
            }
        },
        "id": "3f9f792eb7b6b021c8b9b831931e89373e6b3aa217cfd2da9ef4a852e11bcba6"
    },
[more output]
```

Useful command to more easily scan the CLI output:
```
$ cardano-wallet-jormungandr stake-pool list | jq '.[].metrics | "\(.controlled_stake.quantity) \(.produced_blocks.quantity)"'
Ok.
"4999977900 0"
"0 0"
"259999933700 0"
"0 0"
"297345 0"
"9999989100 0"
"9999933700 0"
"0 0"
"0 0"
"19900 0"
"0 0"
"0 0"
"89999748600 0"
"999999977900 6"
"290000049896 0"
```

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 1]
```diff
@@ -1187,3 +1197,23 @@ instance LiftHandler (Request, ServantErr) where
...
+        ErrListStakePoolsStakeIsUnreachable ->
+            apiError err400 StakeIsUnreachable $ mconcat
+                [ "I can't reach the stake distribution from jörmungandr. "
+                , "Maybe the jörmungandr is not started, or it is on a BFT "
+                , "blockchain instead of, Praos/Genesis?"
+                ]
+        ErrListStakePoolsInternalError _ ->
+            apiError err500 StakeIsUnreachable $ mconcat
+                [ "Internal error."
+                ]
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:1218]: ~StakeIsUnreachable~ --> `UnexpectedError`. 

Also, "Internal error" doesn't tell us anything about what could have happened. What's the case where we would get a 500 here?

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs [hunk 2]
```diff
@@ -471,6 +473,9 @@ isSubrangeOf r1 r2 =
                                   Stake Pools
 -------------------------------------------------------------------------------}
 
+type Stake = Quantity "lovelace" Word64
+type EpochNo = Word64
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/Types.hs:483]: No :no_mouth: ... Don't use type aliases here. Those types are simple enough, if they deserve to exist as types, make them newtypes. Though, `Stake`, has no extra semantic compared to `Quantity "lovelace" Word64`

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs [hunk 3]
```diff
@@ -14,37 +16,169 @@ module Cardano.Wallet.StakePool.Metrics
...
+    { listStakePools
+        :: ExceptT ErrListStakePools m [(PoolId, (Stake, Int))]
+    }
+
+newStakePoolLayer
+    :: NetworkLayer IO tx (BlockHeader, PoolId)
+    -> Trace IO Text
+    -> IO (StakePoolLayer IO)
+newStakePoolLayer nl tr = do
+    mvar <- worker nl tr
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/StakePool/Metrics.hs:105]: Here, we should use: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Registry.hs#L140-L182

These are tested and have all sort of fail-over mechanism for exception and async exception!

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs [hunk 4]
```diff
@@ -14,37 +16,169 @@ module Cardano.Wallet.StakePool.Metrics
...
+    return $ StakePoolLayer
+        { listStakePools = Map.toList <$>
+            combineMetrics
+                (withE1 $ stakeDistribution nl)
+                (liftIO $ readMVar mvar)
+        }
+  where
+    withE1 = withExceptT $ const ()
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/StakePool/Metrics.hs]: So we are discarding an error here :thinking: ?

File: lib/http-bridge/src/Cardano/Wallet/HttpBridge/Network.hs [hunk 5]
```diff
@@ -187,6 +187,8 @@ mkNetworkLayer httpBridge = NetworkLayer
         postSignedTx httpBridge
     , staticBlockchainParameters =
         (block0, byronBlockchainParameters @n)
+    , stakeDistribution =
+        error "stakeDistribution is unimplemented for http-bridge"
```
Comment (KtorZ) [inline lib/http-bridge/src/Cardano/Wallet/HttpBridge/Network.hs:191]: :+1:

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs [hunk 6]
```diff
@@ -116,27 +120,48 @@ serveWallet (cfg, sb, tr) databaseDir listen lj beforeMainLoop = do
...
                     & setBeforeMainLoop (beforeMainLoop (Port wPort) nPort bp)
             let ipcServer = daedalusIPC tracerIPC wPort
-            let apiServer = Server.start settings tracerApi socket wallet
+            let apiServer = Server.start settings tracerApi socket wallet spl
             race_ ipcServer apiServer
 
+    -- Will be used to connect "Cardano.Wallet.StakePool.Metrics" with
+    -- our networkLayer.
+    toSPBlock :: Binary.Block -> (BlockHeader, PoolId)
+    toSPBlock b =
```
Comment (KtorZ) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs]: This function should probably live in the Binary module.

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs [hunk 7]
```diff
@@ -116,27 +120,48 @@ serveWallet (cfg, sb, tr) databaseDir listen lj beforeMainLoop = do
...
+        convertHeader :: Binary.BlockHeader -> BlockHeader
+        convertHeader h = BlockHeader
+            (Binary.slot h)
+            (Quantity $ fromIntegral $ Binary.chainLength h)
+            (Binary.parentHeaderHash h)
+        toJust (Just x) = x
+        toJust Nothing = error
+            "Expected blockheader to contain the id of the\
+            \producing pool, which only happens on Praos/Genesis-blockchains\
+            \for blocks after block0."
```
Comment (Anviking) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs]: I think we should keep this for now and see how things play out with the Sqlite model, instead of adding a `Maybe` in StakePool.Metrics now.

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs [hunk 8]
```diff
@@ -116,27 +120,48 @@ serveWallet (cfg, sb, tr) databaseDir listen lj beforeMainLoop = do
...
-            apiLayer tr nl >>= startServer tr nPort nl
+            apiLayer tr (toWLBlock <$> nl) >>= startServer tr nPort nl
             pure ExitSuccess
         Left e -> handleNetworkStartupError e
   where
     startServer
         :: Trace IO Text
         -> Port "node"
-        -> NetworkLayer IO t (Block Tx)
+        -> NetworkLayer IO t Binary.Block
```
Comment (KtorZ) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs]: We use to call `Binary.Block` -> `J.Block` everywhere else, so let's stick with that :+1:

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs (hunk 7 above)
Comment (KtorZ) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs]: This is still quite unsafe here. So, fetching the genesis block will throw hard in the stake pool network layer. That's bad, because we might actually fetch it more than once in the chain following algorithm! I'd stick with a `Maybe PoolId` at this stage, and do the conversion from `Maybe PoolID` to `PoolId` later ... and safely.

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs [hunk 9]
```diff
@@ -14,37 +16,169 @@ module Cardano.Wallet.StakePool.Metrics
...
+    }
+
+newStakePoolLayer
+    :: NetworkLayer IO tx (BlockHeader, PoolId)
+    -> Trace IO Text
+    -> IO (StakePoolLayer IO)
+newStakePoolLayer nl tr = do
+    mvar <- worker nl tr
+    return $ StakePoolLayer
+        { listStakePools = Map.toList <$>
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/StakePool/Metrics.hs:107]: I don't think we need any particular layer at this stage, a top level function which expects a `NetworkLayer` as argument should be sufficient, no?

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs [hunk 10]
```diff
@@ -14,37 +16,169 @@ module Cardano.Wallet.StakePool.Metrics
...
+    :: forall m readStateErr stakeDistrErr. Monad m
+    => ExceptT stakeDistrErr m (EpochNo, Map PoolId Stake)
+    -> ExceptT readStateErr m State
+    -> ExceptT ErrListStakePools m (Map PoolId (Stake, Int))
+combineMetrics getStakeDistr getActivityState = do
+    (epoch, distr) <- withExceptT (const ErrListStakePoolsStakeIsUnreachable)
+        getStakeDistr
+
+    s <- withInternalErr ErrListStakePoolInternalErrReadStateErr
+        getActivityState
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/StakePool/Metrics.hs]: This is just reading the content of an MVar, isn't? There's no need for an `ExceptT` wrapper here and an `InternalErr`, this can't just fail. Or am I missing something?

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs [hunk 11]
```diff
@@ -14,37 +16,169 @@ module Cardano.Wallet.StakePool.Metrics
...
+
+    m <- case activityForEpoch epoch s of
+        Just x -> return x
+        Nothing -> throwE ErrListStakePoolsMetricsIsUnsynced
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/StakePool/Metrics.hs]: What about returning an empty map? I don't think the error is needed.

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs [hunk 12]
```diff
@@ -14,37 +16,169 @@ module Cardano.Wallet.StakePool.Metrics
...
+    mergeA
+        stakeButNoActivity
+        activityButNoStake
+        (zipWithMatched (\_k stake act -> (stake, act)))
+        distr m
+  where
+    -- What to do when we only find a pool in one of the two data-sources:
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/StakePool/Metrics.hs:139]: The source of truth should be our chain following from the blocks. On top of that, we try to attach the stake. If the stake is unknown, then we should report it as "unknown". This implies some changes to the API :+1:

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs (hunk 7 above)
Comment (KtorZ) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs]: But it will throw when fetching the genesis block. That's not right :grimacing: ...

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs (hunk 3 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/StakePool/Metrics.hs:105]: To be more specific, we don't need any registry here since there's only one global worker. The worker resource here would just be the network layer. There's no preparatory task and, the main task does the chain following.

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs (hunk 9 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/StakePool/Metrics.hs:107]: So, okay, we'll also need to have a connection to the database, and probably some logger metrics as well. so it makes sense to make it part of a bigger object.

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs (hunk 10 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/StakePool/Metrics.hs]: I think you're right, no DB will be able to fail.

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs (hunk 11 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/StakePool/Metrics.hs]: That doesn't sound right to me. With an error, we can report the error like:

```
I can't list stake pools yet because I need to scan the blockchain for metrics first. I'm at 1%
```

File: lib/core/src/Cardano/Wallet/StakePool/Metrics.hs (hunk 9 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/StakePool/Metrics.hs:107]: > So, okay, we'll also need to have a connection to the database, and probably some logger metrics as well. so it makes sense to make it part of a bigger object.

Yes. It is one self-contained thing that can easily be passed around in `Api.Server` and `.Jormungandr.hs`.

Maybe if we separated the `Servant`-serving of the listStakePools-endpoint `StakePoolLayer` wouldn't be needed… But I'm not convinced enough that it will lead to anywhere good.

File: lib/core/src/Cardano/Wallet/Network.hs [hunk 13]
```diff
@@ -287,3 +304,34 @@ follow nl tr cps yield rollback header =
...
+
+data ErrInconsistentTips
+    = ErrInconsistentTips
+    | ErrTipNotFound ErrNetworkTip
+
+atomically
+    :: forall m a t block e. Monad m
+    => (BlockHeader -> BlockHeader -> Bool)
+    -- ^ Compare the preceding, and succeding tip, and judge whether the result
+    -- of the action can be trusted.
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Network.hs:319]: I'd actually remove that parameter. There's no need to provide any particular predicate other then `==`, so we can cut off some indirection / complexity here :+1:

File: lib/core/src/Cardano/Wallet/Network.hs [hunk 14]
```diff
@@ -287,3 +304,34 @@ follow nl tr cps yield rollback header =
                         "Failed to roll backward: " <> show e
                 Right () -> do
                     step cursor'
+
+onSameTip :: BlockHeader -> BlockHeader -> Bool
+onSameTip = (==)
+
+data ErrInconsistentTips
+    = ErrInconsistentTips
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Network.hs:312]: The type name should rather be `ErrAtomically` here. Although, we should be careful about calling this `atomically`.  Things are not atomic here, but, so long as we do read-only calls, there shouldn't be any issue. Any mutations happening in the `ExceptT` will actually happen, and might happen more than once.

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs [hunk 15]
```diff
@@ -1053,6 +1059,14 @@ syncProgress epochLength tip slotNow =
         Quantity $ toEnum $ fromIntegral $
             (100 * bhTip) `div` (bhTip + n1 - n0)
 
+syncProgressRelativeToCurrentTime
+    :: SlotParameters
+    -> BlockHeader
+    -> IO (Quantity "percent" Percentage)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/Types.hs:1065]: :scream: ... `IO` in the `Primitive.Types` module!

More seriously, the time should really be a parameter here. Let's not have effects down here. We've kept it pure until now :) ... Also, shouldn't this be in another PR?

Aaaand.... we've lost the comment that was giving a justification for this unsafe pattern matching :/ ... That's kinda bad.

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 2 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs:483]: I find the `Quantity "lovelace" Word64` a bit clunky and to contain be mostly un-interesting information.

Readability and formatting gets more difficult. For example this fits into one line with the alias
```
        => WhenMissing (ExceptT ErrListStakePools m) PoolId Stake (Stake, Int)

```
but not without:
```
    stakeButNoActivity
        :: Monad m
        => WhenMissing
            (ExceptT ErrListStakePools m)
            PoolId
            (Quantity "lovelace" Word64)
            (Quantity "lovelace" Word64, Int)
```
😥 

I wonder if we can remove some `Quantity "lovelace"` in favour of `Lovelace deriving X via Quantity "lovelace"` eventually…

Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs:483]: The `Quantity` is really just to avoid carrying around naked types like `Int`, or `Word32` and instead, ease type signatures readability. We also use it extensively in the API to disambiguate JSON representation of quantified units. 

Yet, sometimes it makes sense to define a newtype if one really wants to capture a particular meaning (for examples: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Primitive/Types.hs#L1179-L1203). A `Stake` type could work but then, it becomes semantically awkward to me as we are basically saying "this is not any lovelace amount, it's a stake so it's kinda special", whereas, it's not. But, if you think having a dedicated type for this makes more sense, okay. The `Quantity` is really just there so that we don't have to create newtype all the time when what we just want to clarify the meaning given to some raw values (precisely like `Hash` for raw `ByteString`).

Having said that... The argument  "I can't indent it in one line" is pretty weak to justify a new type :grimacing:

**Comment by KtorZ** [comment]
closing in favor of recent PRs and tomorrow's work.
